baseapp/abci_utils.go

```diff
@@ -227,8 +229,38 @@ func (h *DefaultProposalHandler) PrepareProposalHandler() sdk.PrepareProposalHan
 		}
 
 		iterator := h.mempool.Select(ctx, req.Txs)
+		selectedTxsSignersSeqs := make(map[string]uint64)
+		var selectedTxsNums int
 		for iterator != nil {
 			memTx := iterator.Tx()
+			signerData, err := h.signerExtAdapter.GetSigners(memTx)
+			if err != nil {
+				return nil, err
+			}
+			// if the signers aren't in selectedTxsSignersSeqs then we haven't seen them before so we add them and return true
+			// so this tx gets selected.
+			shouldAdd := true
+			txSignersSeqs := make(map[string]uint64)
+			for _, signer := range signerData {
+				seq, ok := selectedTxsSignersSeqs[signer.Signer.String()]
+				if !ok {
+					txSignersSeqs[signer.Signer.String()] = signer.Sequence
+					continue
+				}
+
+				// if we have seen this signer before we check if the sequence we just got is seq+1 and if it is we update the
+				// sequence and return true so this tx gets selected. If it isn't seq+1 we return false so this tx doesn't get
+				// selected (it could be the same sequence or seq+2 which are invalid).
```

Comment:
```suggestion
				// If we have seen this signer before we check if the sequence we just got is
				// seq+1 and if it is we update the sequence and return true so this tx gets
				// selected. If it isn't seq+1 we return false so this tx doesn't get
				// selected (it could be the same sequence or seq+2 which are invalid).
```